feat: add accessor arrays support to `blas/ext/base/gnannsumkbn`

headlessNode opened the pull request:
## Description

> What is the purpose of this pull request?

This pull request:

-   add accessor arrays support to `blas/ext/base/gnannsumkbn`

## Related Issues

> Does this pull request have any related issues?

No. 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

headlessNode committed:
feat: add accessor arrays support

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/gnannsumkbn/docs/types/test.ts:
```diff
@@ -26,7 +27,8 @@ import gnannsumkbn = require( './index' );
 
-	gnannsumkbn( x.length, x, 1, out, 1 ); // $ExpectType NumericArray
+	gnannsumkbn( x.length, x, 1, out, 1 ); // $ExpectType Float64Array
+	gnannsumkbn( x.length, new AccessorArray( x ), 1, new AccessorArray( out ), 1 ); // $ExpectType AccessorArray<number>
```
The `AccessorArray<number>` was suggested by a lint error on local.

kgryte replied:
Yes, that makes sense.

kgryte replied:
Likely applicable elsewhere, as well.

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/gnannsumkbn/docs/types/test.ts:
```diff
@@ -123,7 +125,8 @@ import gnannsumkbn = require( './index' );
 
-	gnannsumkbn.ndarray( x.length, x, 1, 0, out, 1, 0 ); // $ExpectType NumericArray
+	gnannsumkbn.ndarray( x.length, x, 1, 0, out, 1, 0 ); // $ExpectType Float64Array
+	gnannsumkbn.ndarray( x.length, new AccessorArray( x ), 1, 0, new AccessorArray( out ), 1, 0 ); // $ExpectType AccessorArray<number>
```
same here.

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/gnannsumkbn/lib/accessors.js:
```diff
@@ -0,0 +1,121 @@
+	sum = 0.0;
+	ix = offsetX;
+	if ( strideX === 0 ) {
+		if ( isnan( xget( xbuf, ix ) ) ) {
+			oset( obuf, offsetOut, sum );
+			oset( obuf, offsetOut+strideOut, 0 );
+			return out;
+		}
+		oset( obuf, offsetOut, xget( xbuf, ix ) * N );
```
```suggestion
		v = xget( xbuf, ix );
		if ( isnan( v ) ) {
			oset( obuf, offsetOut, sum );
			oset( obuf, offsetOut+strideOut, 0 );
			return out;
		}
		oset( obuf, offsetOut, v * N );
```

kgryte committed:
refactor: avoid duplicate access

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/gnannsumkbn/lib/ndarray.js:
```diff
@@ -68,6 +72,11 @@ function gnannsumkbn( N, x, strideX, offsetX, out, strideOut, offsetOut ) {
+	ox = arraylike2object( x );
+	oo = arraylike2object( out );
+	if ( ox.accessorProtocol || oo.accessorProtocol ) {
+		return accessors( N, ox, strideX, offsetX, oo, strideOut, offsetOut );
```
```suggestion
		accessors( N, ox, strideX, offsetX, oo, strideOut, offsetOut );
		return out;
```

Return original output array, not the object.

kgryte committed:
fix: ensure correct return value

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/gnannsumkbn/test/test.main.js:
```diff
@@ -119,6 +120,100 @@ tape( 'the function calculates the sum of strided array elements (ignoring NaN v
+	v = gnannsumkbn( x.length, x, 1, out, 1 );
+
+	expected = [ 3.0, 9.0 ];
+	t.strictEqual( v.data, out, 'returns expected value' );
```
This isn't correct. `v` should strictly equal `out`.

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/gnannsumkbn/test/test.main.js:
```diff
@@ -187,6 +282,37 @@ tape( 'the function supports `stride` parameters', function test( t ) {
+	v = gnannsumkbn( 5, x, 2, out, 2 );
+
+	expected = [ 5.0, 0.0, 4.0, 0.0 ];
+	t.strictEqual( v.data, out, 'returns expected value' );
```
Same comment throughout this package.

headlessNode committed:
fix: incorrect assertion

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

kgryte reviewed (approved):
LGTM
